# CAPSTONE Project - Neighborhood Clustering

### Importing the dataset - IBM Watson

In [2]:

import types
import pandas as pd
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# @hidden_cell
# The following code accesses a file in your IBM Cloud Object Storage. It includes your credentials.
# You might want to remove those credentials before you share your notebook.
client_a37a82f472884de3af85202afee14c1c = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='umibtusPlqLR-EnYKJqqlcguAERerzijYa3ae_mtuLKO',
    ibm_auth_endpoint="https://iam.bluemix.net/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url='https://s3-api.us-geo.objectstorage.service.networklayer.com')

body = client_a37a82f472884de3af85202afee14c1c.get_object(Bucket='capstonetoronto-donotdelete-pr-6kve0jrbutzhnh',Key='Capstone_project.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_1 = pd.read_csv(body)
df_data_1.head()



Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

### The folloiwng is the shape of the dataset

In [3]:
df_data_1.shape

(288, 3)

### Examining how may boroughs have "Not Assigned" status

In [4]:
df_data_1.Borough.value_counts()

Not assigned        77
Etobicoke           45
North York          38
Scarborough         37
Downtown Toronto    37
Central Toronto     17
West Toronto        13
York                 9
East Toronto         7
East York            6
Mississauga          1
Queen's Park         1
Name: Borough, dtype: int64

#### In the Dataset, 77 entries of borroughs have "Not Assigned Status"

### Since the boroughs which have "Not assgined" status needs to be ignored, I need to drop the same from the dataset. For this, the indexes of such "Not assigned" status is found out. There should be total 77 such entries.

In [5]:
indexnames = df_data_1[df_data_1['Borough'] == "Not assigned"].index
indexnames

Int64Index([  0,   1,   9,  13,  20,  21,  30,  36,  37,  45,  46,  50,  51,
             52,  54,  55,  59,  60,  61,  73,  74,  75,  88,  89,  90, 104,
            105, 106, 120, 121, 136, 137, 148, 149, 155, 161, 162, 167, 175,
            181, 182, 188, 189, 190, 194, 195, 201, 202, 203, 204, 209, 210,
            223, 224, 237, 238, 241, 242, 247, 248, 253, 254, 258, 259, 260,
            261, 263, 264, 274, 275, 276, 277, 278, 279, 280, 281, 287],
           dtype='int64')

### We drop these entries where Boroughs have "Not assigned" status rom the dataset. Since original dataset had 288 rows and 77 of them were deleted, there would be 211 rows. We check the same

In [6]:
df_data_1.drop(indexnames, axis=0, inplace = True)
df_data_1.shape

(211, 3)

### We once again check if our exercise have been succesful.

In [7]:
df_data_1.Borough.value_counts()

Etobicoke           45
North York          38
Scarborough         37
Downtown Toronto    37
Central Toronto     17
West Toronto        13
York                 9
East Toronto         7
East York            6
Mississauga          1
Queen's Park         1
Name: Borough, dtype: int64

#### There are no "Not Assigned" Boroughs in the resulting dataset.

### Next, the number of unique Postcodes was found.

In [8]:
print("Number of Unique Postal codes is :", len(df_data_1.Postcode.unique()))

Number of Unique Postal codes is : 103


### Next, we check if there are any neighbourhood/s which have "Not Assigned" Status

In [9]:
df_data_1[df_data_1.Neighbourhood =='Not assigned']

Postcode       Borough Neighbourhood
8      M7A  Queen's Park  Not assigned

#### There is one entry where Neighbourhood has "Not assigned" status. We need to replace the Neighbourhood value with Borough name, which we do using the below code.

In [10]:
df_data_1.Neighbourhood.replace("Not assigned", "Queen's Park", inplace = True)

#### We check whether the same has been replaced.

In [11]:
df_data_1[df_data_1['Postcode'] == 'M7A']

Postcode       Borough Neighbourhood
8      M7A  Queen's Park  Queen's Park

### The neighbourhoods are joined using the below mentioned code using group by and lambda. It is saved in a different dataframe

In [12]:
df= pd.DataFrame(df_data_1.groupby("Postcode")['Neighbourhood'].apply(lambda tags: ','.join(tags)))
df

Neighbourhood
Postcode                                                   
M1B                                           Rouge,Malvern
M1C                    Highland Creek,Rouge Hill,Port Union
M1E                         Guildwood,Morningside,West Hill
M1G                                                  Woburn
M1H                                               Cedarbrae
M1J                                     Scarborough Village
M1K               East Birchmount Park,Ionview,Kennedy Park
M1L                           Clairlea,Golden Mile,Oakridge
M1M           Cliffcrest,Cliffside,Scarborough Village West
M1N                              Birch Cliff,Cliffside West
M1P       Dorset Park,Scarborough Town Centre,Wexford He...
M1R                                        Maryvale,Wexford
M1S                                               Agincourt
M1T                   Clarks Corners,Sullivan,Tam O'Shanter
M1V       Agincourt North,L'Amoreaux East,Milliken,Steel...
M1W                                         L'Amoreaux West
M1X                                             Upper Rouge
M2H                                       Hillcrest Village
M2J                              Fairview,Henry Farm,Oriole
M2K                                         Bayview Village
M2L                                 Silver Hills,York Mills
M2M                                  Newtonbrook,Willowdale
M2N                                        Willowdale South
M2P                                         York Mills West
M2R                                         Willowdale West
M3A                                               Parkwoods
M3B                                         Don Mills North
M3C                         Flemingdon Park,Don Mills South
M3H           Bathurst Manor,Downsview North,Wilson Heights
M3J                          Northwood Park,York University
...                                                     ...
M6C                                      Humewood-Cedarvale
M6E                                     Caledonia-Fairbanks
M6G                                                Christie
M6H                             Dovercourt Village,Dufferin
M6J                                 Little Portugal,Trinity
M6K              Brockton,Exhibition Place,Parkdale Village
M6L                        Downsview,North Park,Upwood Park
M6M             Del Ray,Keelesdale,Mount Dennis,Silverthorn
M6N                            The Junction North,Runnymede
M6P                            High Park,The Junction South
M6R                                   Parkdale,Roncesvalles
M6S                                       Runnymede,Swansea
M7A                                            Queen's Park
M7R                   Canada Post Gateway Processing Centre
M7Y       Business Reply Mail Processing Centre 969 Eastern
M8V              Humber Bay Shores,Mimico South,New Toronto
M8W                                   Alderwood,Long Branch
M8X             The Kingsway,Montgomery Road,Old Mill North
M8Y       Humber Bay,King's Mill Park,Kingsway Park Sout...
M8Z       Kingsway Park South West,Mimico NW,The Queensw...
M9A                                        Islington Avenue
M9B       Cloverdale,Islington,Martin Grove,Princess Gar...
M9C       Bloordale Gardens,Eringate,Markland Wood,Old B...
M9L                                           Humber Summit
M9M                                         Emery,Humberlea
M9N                                                  Weston
M9P                                               Westmount
M9R       Kingsview Village,Martin Grove Gardens,Richvie...
M9V       Albion Gardens,Beaumond Heights,Humbergate,Jam...
M9W                                               Northwest

[103 rows x 1 columns]

### Groupby statment results in indexing the Key. The same can be seen as below.

In [13]:
df.head(5)

Neighbourhood
Postcode                                      
M1B                              Rouge,Malvern
M1C       Highland Creek,Rouge Hill,Port Union
M1E            Guildwood,Morningside,West Hill
M1G                                     Woburn
M1H                                  Cedarbrae

### The indexing is removed below.

In [14]:
df.reset_index(inplace = True)
df.head(5)

Postcode                         Neighbourhood
0      M1B                         Rouge,Malvern
1      M1C  Highland Creek,Rouge Hill,Port Union
2      M1E       Guildwood,Morningside,West Hill
3      M1G                                Woburn
4      M1H                             Cedarbrae

### Next, we need to merge the 2 dataframes on Postcode for merging Neighbourhood. Hnece first, we drop the "Neighbourhood column from the original dataframe.

In [15]:
df_data_1.drop('Neighbourhood', axis= 1, inplace =True)
df_data_1.head(5)

Postcode           Borough
2      M3A        North York
3      M4A        North York
4      M5A  Downtown Toronto
5      M5A  Downtown Toronto
6      M6A        North York

### Merging the two dataframes.

In [16]:
df_final = pd.merge(df_data_1, df, on= 'Postcode')
df_final

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M6A        North York   
5        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
9        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
12       M4B         East York   
13       M5B  Downtown Toronto   
14       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M9B         Etobicoke   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M1C       Scarborough   
24       M3C        North York   
25       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
..       ...               ...   
181      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
186      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
190      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
192      M5X  Downtown Toronto   
193      M8X         Etobicoke   
194      M8X         Etobicoke   
195      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
199      M8Y         Etobicoke   
200      M8Y         Etobicoke   
201      M8Y         Etobicoke   
202      M8Y         Etobicoke   
203      M8Y         Etobicoke   
204      M8Y         Etobicoke   
205      M8Y         Etobicoke   
206      M8Z         Etobicoke   
207      M8Z         Etobicoke   
208      M8Z         Etobicoke   
209      M8Z         Etobicoke   
210      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                             Harbourfront,Regent Park  
3                             Harbourfront,Regent Park  
4                      Lawrence Heights,Lawrence Manor  
5                      Lawrence Heights,Lawrence Manor  
6                                         Queen's Park  
7                                     Islington Avenue  
8                                        Rouge,Malvern  
9                                        Rouge,Malvern  
10                                     Don Mills North  
11                      Woodbine Gardens,Parkview Hill  
12                      Woodbine Gardens,Parkview Hill  
13                             Ryerson,Garden District  
14                             Ryerson,Garden District  
15                                           Glencairn  
16   Cloverdale,Islington,Martin Grove,Princess Gar...  
17   Cloverdale,Islington,Martin Grove,Princess Gar...  
18   Cloverdale,Islington,Martin Grove,Princess Gar...  
19   Cloverdale,Islington,Martin Grove,Princess Gar...  
20   Cloverdale,Islington,Martin Grove,Princess Gar...  
21                Highland Creek,Rouge Hill,Port Union  
22                Highland Creek,Rouge Hill,Port Union  
23                Highland Creek,Rouge Hill,Port Union  
24                     Flemingdon Park,Don Mills South  
25                     Flemingdon Park,Don Mills South  
26                                    Woodbine Heights  
27                                      St. James Town  
28                                  Humewood-Cedarvale  
29   Bloordale Gardens,Eringate,Markland Wood,Old B...  
..                                                 ...  
181  Albion Gardens,Beaumond Heights,Humbergate,Jam...  
182           

### Sicne there were 211 rows in the original dataframe, the resulting merged dataframe will also contain 211 rows, as shown below. 

In [17]:
df_final.shape

(211, 3)

### However, this dataframe contains many duplicate values since the number of unqiue postcodes is only 103. We drop the duplicates ans check the final shape, which should contain 103 rows.

In [18]:
df_final.drop_duplicates(inplace = True)
df_final.shape

(103, 3)

### resetting the index

In [19]:
df_final.reset_index(inplace =True)
df_final

index Postcode           Borough  \
0        0      M3A        North York   
1        1      M4A        North York   
2        2      M5A  Downtown Toronto   
3        4      M6A        North York   
4        6      M7A      Queen's Park   
5        7      M9A         Etobicoke   
6        8      M1B       Scarborough   
7       10      M3B        North York   
8       11      M4B         East York   
9       13      M5B  Downtown Toronto   
10      15      M6B        North York   
11      16      M9B         Etobicoke   
12      21      M1C       Scarborough   
13      24      M3C        North York   
14      26      M4C         East York   
15      27      M5C  Downtown Toronto   
16      28      M6C              York   
17      29      M9C         Etobicoke   
18      33      M1E       Scarborough   
19      36      M4E      East Toronto   
20      37      M5E  Downtown Toronto   
21      38      M6E              York   
22      39      M1G       Scarborough   
23      40      M4G         East York   
24      41      M5G  Downtown Toronto   
25      42      M6G  Downtown Toronto   
26      43      M1H       Scarborough   
27      44      M2H        North York   
28      45      M3H        North York   
29      48      M4H         East York   
..     ...      ...               ...   
73     130      M4R   Central Toronto   
74     131      M5R   Central Toronto   
75     134      M6R      West Toronto   
76     136      M7R       Mississauga   
77     137      M9R         Etobicoke   
78     141      M1S       Scarborough   
79     142      M4S   Central Toronto   
80     143      M5S  Downtown Toronto   
81     145      M6S      West Toronto   
82     147      M1T       Scarborough   
83     150      M4T   Central Toronto   
84     152      M5T  Downtown Toronto   
85     155      M1V       Scarborough   
86     159      M4V   Central Toronto   
87     164      M5V  Downtown Toronto   
88     171      M8V         Etobicoke   
89     174      M9V         Etobicoke   
90     182      M1W       Scarborough   
91     183      M4W  Downtown Toronto   
92     184      M5W  Downtown Toronto   
93     185      M8W         Etobicoke   
94     187      M9W         Etobicoke   
95     188      M1X       Scarborough   
96     189      M4X  Downtown Toronto   
97     191      M5X  Downtown Toronto   
98     193      M8X         Etobicoke   
99     196      M4Y  Downtown Toronto   
100    197      M7Y      East Toronto   
101    198      M8Y         Etobicoke   
102    206      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                             Harbourfront,Regent Park  
3                      Lawrence Heights,Lawrence Manor  
4                                         Queen's Park  
5                                     Islington Avenue  
6                                        Rouge,Malvern  
7                                      Don Mills North  
8                       Woodbine Gardens,Parkview Hill  
9                              Ryerson,Garden District  
10                                           Glencairn  
11   Cloverdale,Islington,Martin Grove,Princess Gar...  
12                Highland Creek,Rouge Hill,Port Union  
13                     Flemingdon Park,Don Mills South  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Bloordale Gardens,Eringate,Markland Wood,Old B...  
18                     Guildwood,Morningside,West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                                 Caledonia-Fairbanks  
22                                              Woburn  
23                                             Leaside  
24                                  C

### Removing the index column,

In [20]:
df_final.drop('index', axis=1, inplace = True)
df_final

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A      Queen's Park   
5        M9A         Etobicoke   
6        M1B       Scarborough   
7        M3B        North York   
8        M4B         East York   
9        M5B  Downtown Toronto   
10       M6B        North York   
11       M9B         Etobicoke   
12       M1C       Scarborough   
13       M3C        North York   
14       M4C         East York   
15       M5C  Downtown Toronto   
16       M6C              York   
17       M9C         Etobicoke   
18       M1E       Scarborough   
19       M4E      East Toronto   
20       M5E  Downtown Toronto   
21       M6E              York   
22       M1G       Scarborough   
23       M4G         East York   
24       M5G  Downtown Toronto   
25       M6G  Downtown Toronto   
26       M1H       Scarborough   
27       M2H        North York   
28       M3H        North York   
29       M4H         East York   
..       ...               ...   
73       M4R   Central Toronto   
74       M5R   Central Toronto   
75       M6R      West Toronto   
76       M7R       Mississauga   
77       M9R         Etobicoke   
78       M1S       Scarborough   
79       M4S   Central Toronto   
80       M5S  Downtown Toronto   
81       M6S      West Toronto   
82       M1T       Scarborough   
83       M4T   Central Toronto   
84       M5T  Downtown Toronto   
85       M1V       Scarborough   
86       M4V   Central Toronto   
87       M5V  Downtown Toronto   
88       M8V         Etobicoke   
89       M9V         Etobicoke   
90       M1W       Scarborough   
91       M4W  Downtown Toronto   
92       M5W  Downtown Toronto   
93       M8W         Etobicoke   
94       M9W         Etobicoke   
95       M1X       Scarborough   
96       M4X  Downtown Toronto   
97       M5X  Downtown Toronto   
98       M8X         Etobicoke   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   
101      M8Y         Etobicoke   
102      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                             Harbourfront,Regent Park  
3                      Lawrence Heights,Lawrence Manor  
4                                         Queen's Park  
5                                     Islington Avenue  
6                                        Rouge,Malvern  
7                                      Don Mills North  
8                       Woodbine Gardens,Parkview Hill  
9                              Ryerson,Garden District  
10                                           Glencairn  
11   Cloverdale,Islington,Martin Grove,Princess Gar...  
12                Highland Creek,Rouge Hill,Port Union  
13                     Flemingdon Park,Don Mills South  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Bloordale Gardens,Eringate,Markland Wood,Old B...  
18                     Guildwood,Morningside,West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                                 Caledonia-Fairbanks  
22                                              Woburn  
23                                             Leaside  
24                                  Central Bay Street  
25                                            Christie  
26                                           Cedarbrae  
27                                   Hillcrest Village  
28       Bathurst Manor,Downsview North,Wilson Heights  
29                                    Thorncliffe Park  
..                                                 ...  
73                                  North Toronto West  
74            

## The final shape of the dataframe:

In [109]:
df_final.shape

(103, 3)

# The final dataframe consists of 103 rows and 3 columns

###  

# Clustering and exploring neighbourhood

### Loading the dataset. The same is done using the Geospatial_Cordinates.csv provided by the link in the "My Submission" Page

In [21]:

body = client_a37a82f472884de3af85202afee14c1c.get_object(Bucket='capstonetoronto-donotdelete-pr-6kve0jrbutzhnh',Key='Geospatial_Coordinates.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_3 = pd.read_csv(body)
df_data_3.head()



Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Finding the shape of the dataset. 

In [22]:
df_data_3.shape

(103, 3)

### Sicne the two datasets have to be merged on Postcode- the key, we rename the column in the new dataframe by Postcode.

In [23]:
df_data_3 = df_data_3.rename(columns = {"Postal Code": "Postcode"})
df_data_3.head(5)

Postcode   Latitude  Longitude
0      M1B  43.806686 -79.194353
1      M1C  43.784535 -79.160497
2      M1E  43.763573 -79.188711
3      M1G  43.770992 -79.216917
4      M1H  43.773136 -79.239476

### Merging the two dataframes

In [24]:
df_latlon = pd.merge(df_final,df_data_3,on='Postcode')
df_latlon.head(5)

Postcode           Borough                    Neighbourhood   Latitude  \
0      M3A        North York                        Parkwoods  43.753259   
1      M4A        North York                 Victoria Village  43.725882   
2      M5A  Downtown Toronto         Harbourfront,Regent Park  43.654260   
3      M6A        North York  Lawrence Heights,Lawrence Manor  43.718518   
4      M7A      Queen's Park                     Queen's Park  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

#### Using geopy library to get the latitude and longitude values of Toronto.

In [25]:
from geopy.geocoders import Nominatim

In [26]:
address = 'Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [27]:
!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/DSX-Python35

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    folium-0.5.0               |             py_0          45 KB  conda-forge
    certifi-2018.8.24          |        py35_1001         139 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    ca-certificates-2019.3.9   |       hecc5488_0         146 KB  conda-forge
    openssl-1.0.2r             |       h14c3975_0         3.1 MB  conda-forge
    altair-2.2.2               |           py35_1         462 KB  conda-forge
    branca-0.3.1               |             py_0          25 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         4.0 MB

The following NEW packages will

### Creating a map of Toronto with neighborhoods superimposed on top.

In [28]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)


for lat, lng, borough, neighborhood in zip(df_latlon['Latitude'], df_latlon['Longitude'], df_latlon['Borough'], df_latlon['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Checking the number of Boroughs in the dataset.

In [29]:
df_latlon.Borough.value_counts()

North York          24
Downtown Toronto    18
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East Toronto         5
York                 5
East York            5
Mississauga          1
Queen's Park         1
Name: Borough, dtype: int64

### Since North York has the highest number of entries, it is decided to explore North York. A separate dataframe is created for North York only

In [30]:
northyork_data =df_latlon[df_latlon['Borough']=='North York'].reset_index(drop = True)
northyork_data.head(5)

Postcode     Borough                    Neighbourhood   Latitude  Longitude
0      M3A  North York                        Parkwoods  43.753259 -79.329656
1      M4A  North York                 Victoria Village  43.725882 -79.315572
2      M6A  North York  Lawrence Heights,Lawrence Manor  43.718518 -79.464763
3      M3B  North York                  Don Mills North  43.745906 -79.352188
4      M6B  North York                        Glencairn  43.709577 -79.445073

### Removing the Postcode from the dataframe, no longer required.

In [31]:
northyork_data.drop('Postcode', axis= 1, inplace= True)
northyork_data.head(5)

Borough                    Neighbourhood   Latitude  Longitude
0  North York                        Parkwoods  43.753259 -79.329656
1  North York                 Victoria Village  43.725882 -79.315572
2  North York  Lawrence Heights,Lawrence Manor  43.718518 -79.464763
3  North York                  Don Mills North  43.745906 -79.352188
4  North York                        Glencairn  43.709577 -79.445073

### Getting the geograhical coordinates of North York

In [32]:
address = 'North York'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 43.7708175, -79.4132998.


### Creating a map of North York with neighbourhoods in it.

In [33]:
map_northyork = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(northyork_data['Latitude'], northyork_data['Longitude'], northyork_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_northyork)  
    
map_northyork

### Defining Foursquare Credentials and Version

In [34]:
CLIENT_ID = 'TVMEP44S5OUTVYLHCLQLSMI0TF3YD2WBUU2S5AILD55EHUBN' 
CLIENT_SECRET = 'PQMZV25AASDDEW13LH4YSD1O4MCYHDLCC10J4OHKMG2IP0CT' 
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: TVMEP44S5OUTVYLHCLQLSMI0TF3YD2WBUU2S5AILD55EHUBN
CLIENT_SECRET:PQMZV25AASDDEW13LH4YSD1O4MCYHDLCC10J4OHKMG2IP0CT


### Exploring the first neighbourhood in the dataset.

In [35]:
northyork_data.loc[0, 'Neighbourhood']

'Parkwoods'

### Getting the neighbourhoods cordinates

In [36]:
neighbourhood_latitude = northyork_data.loc[0, 'Latitude'] 
neighbourhood_longitude = northyork_data.loc[0, 'Longitude'] 

neighbourhood_name = northyork_data.loc[0, 'Neighbourhood'] 

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


### top 100 venues that are in Parkwoodsl within a radius of 500 meters

In [37]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=TVMEP44S5OUTVYLHCLQLSMI0TF3YD2WBUU2S5AILD55EHUBN&client_secret=PQMZV25AASDDEW13LH4YSD1O4MCYHDLCC10J4OHKMG2IP0CT&v=20180604&ll=43.7532586,-79.3296565&radius=500&limit=100'

### GET request and examining the resutls

In [38]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas datafram

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5cde3e641ed21914c03db060'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4e8d9dcdd5fbbbb6b3003c7b-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/park_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d163941735',
         'name': 'Park',
         'pluralName': 'Parks',
         'primary': True,
         'shortName': 'Park'}],
       'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'location': {'address': 'Toronto',
        'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'distance': 245,
        'formattedAddress': ['Toronto', 'Toronto ON', 'Canada'],
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.751976046055574,
          'lng': -79.33214044722958}],
        'lat

### get_category_type function.

In [39]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

### clean the json and structure it into a pandas dataframe.

In [40]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0  Brookbanks Park                  Park  43.751976 -79.332140
1              KFC  Fast Food Restaurant  43.754387 -79.333021
2    Variety Store     Food & Drink Shop  43.751974 -79.333114
3    GreenWin pool                  Pool  43.756232 -79.333842

In [41]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare.


### Explore Neighborhoods in North York

In [42]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### A new dataframe is created.

In [43]:
northyork_venues = getNearbyVenues(names=northyork_data['Neighbourhood'],
                                   latitudes=northyork_data['Latitude'],
                                   longitudes=northyork_data['Longitude']
                                  )

Parkwoods
Victoria Village
Lawrence Heights,Lawrence Manor
Don Mills North
Glencairn
Flemingdon Park,Don Mills South
Hillcrest Village
Bathurst Manor,Downsview North,Wilson Heights
Fairview,Henry Farm,Oriole
Northwood Park,York University
Bayview Village
CFB Toronto,Downsview East
Silver Hills,York Mills
Downsview West
Downsview,North Park,Upwood Park
Humber Summit
Newtonbrook,Willowdale
Downsview Central
Bedford Park,Lawrence Manor East
Emery,Humberlea
Willowdale South
Downsview Northwest
York Mills West
Willowdale West


### size of the resulting dataframe

In [44]:
print(northyork_venues.shape)
northyork_venues.head()

(243, 7)


Neighborhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3         Parkwoods               43.753259               -79.329656   
4  Victoria Village               43.725882               -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1                     KFC       43.754387       -79.333021   
2           Variety Store       43.751974       -79.333114   
3           GreenWin pool       43.756232       -79.333842   
4  Victoria Village Arena       43.723481       -79.315635   

         Venue Category  
0                  Park  
1  Fast Food Restaurant  
2     Food & Drink Shop  
3                  Pool  
4          Hockey Arena

### venues for each neighborhood

In [45]:
northyork_venues.groupby('Neighborhood').count()

Neighbourhood Latitude  \
Neighborhood                                                            
Bathurst Manor,Downsview North,Wilson Heights                      19   
Bayview Village                                                     4   
Bedford Park,Lawrence Manor East                                   23   
CFB Toronto,Downsview East                                          3   
Don Mills North                                                     4   
Downsview Central                                                   4   
Downsview Northwest                                                 5   
Downsview West                                                      4   
Downsview,North Park,Upwood Park                                    4   
Emery,Humberlea                                                     2   
Fairview,Henry Farm,Oriole                                         66   
Flemingdon Park,Don Mills South                                    21   
Glencairn                                                           4   
Hillcrest Village                                                   6   
Humber Summit                                                       2   
Lawrence Heights,Lawrence Manor                                    15   
Northwood Park,York University                                      5   
Parkwoods                                                           4   
Silver Hills,York Mills                                             2   
Victoria Village                                                    4   
Willowdale South                                                   34   
Willowdale West                                                     5   
York Mills West                                                     3   

                                               Neighbourhood Longitude  Venue  \
Neighborhood                                                                    
Bathurst Manor,Downsview North,Wilson Heights                       19     19   
Bayview Village                                                      4      4   
Bedford Park,Lawrence Manor East                                    23     23   
CFB Toronto,Downsview East                                           3      3   
Don Mills North                                                      4      4   
Downsview Central                                                    4      4   
Downsview Northwest                                                  5      5   
Downsview West                                                       4      4   
Downsview,North Park,Upwood Park                                     4      4   
Emery,Humberlea                                                      2      2   
Fairview,Henry Farm,Oriole                                          66     66   
Flemingdon Park,Don Mills South                                     21     21   
Glencairn                                                            4      4   
Hillcrest Village                                                    6      6   
Humber Summit                                                        2      2   
Lawrence Heights,Lawrence Manor                                     15     15   
Northwood Park,York University                                       5      5   
Parkwoods                                                            4      4   
Silver Hills,York Mills                                              2      2   
Victoria Village                                                     4      4   
Willowdale South                                                    34     34   
Willowdale West                                                      5      5   
York Mills West                                                      3      3   

                                               Venue Latitude  \
Neighborhood                                                    
Bathurst Manor,Downsview North,Wilson Heights              19   


### unique categories in the dataset

In [46]:
print('There are {} uniques categories.'.format(len(northyork_venues['Venue Category'].unique())))

There are 108 uniques categories.


### Analyze Each Neighborhood

In [47]:
northyork_onehot = pd.get_dummies(northyork_venues[['Venue Category']], prefix="", prefix_sep="")


northyork_onehot['Neighborhood'] = northyork_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [northyork_onehot.columns[-1]] + list(northyork_onehot.columns[:-1])
northyork_onehot = northyork_onehot[fixed_columns]

northyork_onehot.head()

Neighborhood  Accessories Store  Airport  American Restaurant  Arcade  \
0         Parkwoods                  0        0                    0       0   
1         Parkwoods                  0        0                    0       0   
2         Parkwoods                  0        0                    0       0   
3         Parkwoods                  0        0                    0       0   
4  Victoria Village                  0        0                    0       0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Bakery  Bank  \
0                    0                 0                   0       0     0   
1                    0                 0                   0       0     0   
2                    0                 0                   0       0     0   
3                    0                 0                   0       0     0   
4                    0                 0                   0       0     0   

       ...        Sushi Restaurant  Tea Room  Thai Restaurant  Theater  \
0      ...                       0         0                0        0   
1      ...                       0         0                0        0   
2      ...                       0         0                0        0   
3      ...                       0         0                0        0   
4      ...                       0         0                0        0   

   Toy / Game Store  Video Game Store  Video Store  Vietnamese Restaurant  \
0                 0                 0            0                      0   
1                 0                 0            0                      0   
2                 0                 0            0                      0   
3                 0                 0            0                      0   
4                 0                 0            0                      0   

   Wings Joint  Women's Store  
0            0              0  
1            0              0  
2            0              0  
3            0              0  
4            0              0  

[5 rows x 109 columns]

### the new dataframe size

In [48]:
northyork_onehot.shape

(243, 109)

### grouping rows by neighborhood by taking the mean of the frequency of occurrence of each category

In [49]:
northyork_grouped = northyork_onehot.groupby('Neighborhood').mean().reset_index()
northyork_grouped

Neighborhood  Accessories Store  \
0   Bathurst Manor,Downsview North,Wilson Heights           0.000000   
1                                 Bayview Village           0.000000   
2                Bedford Park,Lawrence Manor East           0.000000   
3                      CFB Toronto,Downsview East           0.000000   
4                                 Don Mills North           0.000000   
5                               Downsview Central           0.000000   
6                             Downsview Northwest           0.000000   
7                                  Downsview West           0.000000   
8                Downsview,North Park,Upwood Park           0.000000   
9                                 Emery,Humberlea           0.000000   
10                     Fairview,Henry Farm,Oriole           0.000000   
11                Flemingdon Park,Don Mills South           0.000000   
12                                      Glencairn           0.000000   
13                              Hillcrest Village           0.000000   
14                                  Humber Summit           0.000000   
15                Lawrence Heights,Lawrence Manor           0.066667   
16                 Northwood Park,York University           0.000000   
17                                      Parkwoods           0.000000   
18                        Silver Hills,York Mills           0.000000   
19                               Victoria Village           0.000000   
20                               Willowdale South           0.000000   
21                                Willowdale West           0.000000   
22                                York Mills West           0.000000   

     Airport  American Restaurant  Arcade  Arts & Crafts Store  \
0   0.000000             0.000000    0.00             0.000000   
1   0.000000             0.000000    0.00             0.000000   
2   0.000000             0.043478    0.00             0.000000   
3   0.333333             0.000000    0.00             0.000000   
4   0.000000             0.000000    0.00             0.000000   
5   0.000000             0.000000    0.00             0.000000   
6   0.000000             0.000000    0.00             0.000000   
7   0.000000             0.000000    0.00             0.000000   
8   0.000000             0.000000    0.00             0.000000   
9   0.000000             0.000000    0.00             0.000000   
10  0.000000             0.015152    0.00             0.000000   
11  0.000000             0.000000    0.00             0.000000   
12  0.000000             0.000000    0.25             0.000000   
13  0.000000             0.000000    0.00             0.000000   
14  0.000000             0.000000    0.00             0.000000   
15  0.000000             0.000000    0.00             0.066667   
16  0.000000             0.000000    0.00             0.000000   
17  0.000000             0.000000    0.00             0.000000   
18  0.000000             0.000000    0.00             0.000000   
19  0.000000             0.000000    0.00             0.000000   
20  0.000000             0.000000    0.00             0.029412   
21  0.000000             0.000000    0.00             0.000000   
22  0.000000             0.000000    0.00             0.000000   

    Asian Restaurant  Athletics & Sports    Bakery      Bank      ...        \
0           0.000000            0.000000  0.000000  0.052632      ...         
1           0.000000            0.000000  0.000000  0.250000      ...         
2           0.000000            0.000000  0.000000  0.000000      ...         
3           0.000000            0.000000  0.000000  0.000000      ...         
4           0.000000            0.000000  0.000000  0.000000      ...         
5           0.000000            0.000000  0.000000  0.000000      ...         
6           0.000000            0.200000  0.000000  0.000000      ...         
7           0.000000            0.000000  0.000000  0.250000      ...         
8           

### The size of the new dataframe

In [50]:
northyork_grouped.shape

(23, 109)

### Each neighborhood along with the top 5 most common venues

In [51]:
num_top_venues = 5

for hood in northyork_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = northyork_grouped[northyork_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bathurst Manor,Downsview North,Wilson Heights----
              venue  freq
0       Coffee Shop  0.11
1  Sushi Restaurant  0.05
2        Restaurant  0.05
3    Sandwich Place  0.05
4     Shopping Mall  0.05


----Bayview Village----
                 venue  freq
0   Chinese Restaurant  0.25
1                 Café  0.25
2  Japanese Restaurant  0.25
3                 Bank  0.25
4    Accessories Store  0.00


----Bedford Park,Lawrence Manor East----
                  venue  freq
0    Italian Restaurant  0.09
1  Fast Food Restaurant  0.09
2           Coffee Shop  0.09
3         Grocery Store  0.04
4               Butcher  0.04


----CFB Toronto,Downsview East----
         venue  freq
0      Airport  0.33
1         Park  0.33
2  Snack Place  0.33
3     Pharmacy  0.00
4    Pet Store  0.00


----Don Mills North----
                  venue  freq
0  Gym / Fitness Center  0.25
1  Caribbean Restaurant  0.25
2                  Café  0.25
3   Japanese Restaurant  0.25
4        Massage Studio  0.0

### Putting the same into a pandas dataframe

In [52]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### Creating the new dataframe and displaying the top 10 venues for each neighborhood.

In [53]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = northyork_grouped['Neighborhood']

for ind in np.arange(northyork_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(northyork_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0  Bathurst Manor,Downsview North,Wilson Heights           Coffee Shop   
1                                Bayview Village    Chinese Restaurant   
2               Bedford Park,Lawrence Manor East  Fast Food Restaurant   
3                     CFB Toronto,Downsview East           Snack Place   
4                                Don Mills North  Caribbean Restaurant   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0   Fried Chicken Joint         Shopping Mall  Middle Eastern Restaurant   
1                  Café                  Bank        Japanese Restaurant   
2    Italian Restaurant           Coffee Shop           Greek Restaurant   
3               Airport                  Park          Electronics Store   
4  Gym / Fitness Center                  Café        Japanese Restaurant   

  5th Most Common Venue    6th Most Common Venue       7th Most Common Venue  \
0    Frozen Yogurt Shop                Pet Store                    Pharmacy   
1     Electronics Store  Comfort Food Restaurant  Construction & Landscaping   
2     Indian Restaurant                     Café                Liquor Store   
3           Coffee Shop  Comfort Food Restaurant  Construction & Landscaping   
4         Women's Store        Electronics Store  Construction & Landscaping   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Pizza Place         Deli / Bodega            Bridal Shop  
1        Cosmetics Shop         Deli / Bodega       Department Store  
2               Butcher              Pharmacy            Pizza Place  
3        Cosmetics Shop         Deli / Bodega       Department Store  
4        Cosmetics Shop         Deli / Bodega       Department Store

### Run k-means to cluster the neighborhood into 5 clusters.

In [54]:
from sklearn.cluster import KMeans

kclusters = 5

northyork_grouped_clustering = northyork_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(northyork_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 4, 0, 1, 4, 0, 0, 0, 0, 2], dtype=int32)

### Creating a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [55]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)



In [56]:
northyork_merged = northyork_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
northyork_merged = northyork_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

northyork_merged.head() # check the last columns!

Borough                    Neighbourhood   Latitude  Longitude  \
0  North York                        Parkwoods  43.753259 -79.329656   
1  North York                 Victoria Village  43.725882 -79.315572   
2  North York  Lawrence Heights,Lawrence Manor  43.718518 -79.464763   
3  North York                  Don Mills North  43.745906 -79.352188   
4  North York                        Glencairn  43.709577 -79.445073   

   Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0             0.0                  Park                    Pool   
1             0.0           Pizza Place             Coffee Shop   
2             0.0        Clothing Store  Furniture / Home Store   
3             4.0  Caribbean Restaurant    Gym / Fitness Center   
4             0.0                Arcade                  Bakery   

  3rd Most Common Venue  4th Most Common Venue 5th Most Common Venue  \
0     Food & Drink Shop   Fast Food Restaurant        Discount Store   
1          Hockey Arena  Portuguese Restaurant         Women's Store   
2         Women's Store     Miscellaneous Shop   Arts & Crafts Store   
3                  Café    Japanese Restaurant         Women's Store   
4                   Pub    Japanese Restaurant         Women's Store   

  6th Most Common Venue       7th Most Common Venue  \
0        Clothing Store                 Coffee Shop   
1        Discount Store     Comfort Food Restaurant   
2              Boutique                 Coffee Shop   
3     Electronics Store  Construction & Landscaping   
4     Electronics Store     Comfort Food Restaurant   

        8th Most Common Venue       9th Most Common Venue  \
0     Comfort Food Restaurant  Construction & Landscaping   
1  Construction & Landscaping              Cosmetics Shop   
2                 Event Space                   Gift Shop   
3              Cosmetics Shop               Deli / Bodega   
4  Construction & Landscaping              Cosmetics Shop   

  10th Most Common Venue  
0         Cosmetics Shop  
1          Deli / Bodega  
2      Accessories Store  
3       Department Store  
4          Deli / Bodega

### Checking whether there are any NaN entries

In [83]:
northyork_merged.isnull().sum()

Borough                   0
Neighbourhood             0
Latitude                  0
Longitude                 0
Cluster Labels            1
1st Most Common Venue     1
2nd Most Common Venue     1
3rd Most Common Venue     1
4th Most Common Venue     1
5th Most Common Venue     1
6th Most Common Venue     1
7th Most Common Venue     1
8th Most Common Venue     1
9th Most Common Venue     1
10th Most Common Venue    1
dtype: int64

### Seems like there is one entry with Nan value. Checking which entry it is.

In [96]:
northyork_merged[northyork_merged['10th Most Common Venue'].isnull()]

Borough           Neighbourhood   Latitude  Longitude  Cluster Labels  \
16  North York  Newtonbrook,Willowdale  43.789053 -79.408493             NaN   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
16                   NaN                   NaN                   NaN   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
16                   NaN                   NaN                   NaN   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
16                   NaN                   NaN                   NaN   

   10th Most Common Venue  
16                    NaN

### Dropping the NaN netity from the dataset.

In [101]:
northyork_merged.dropna(axis =0, inplace = True)
northyork_merged.isnull().sum()

Borough                   0
Neighbourhood             0
Latitude                  0
Longitude                 0
Cluster Labels            0
1st Most Common Venue     0
2nd Most Common Venue     0
3rd Most Common Venue     0
4th Most Common Venue     0
5th Most Common Venue     0
6th Most Common Venue     0
7th Most Common Venue     0
8th Most Common Venue     0
9th Most Common Venue     0
10th Most Common Venue    0
dtype: int64

### Visualizing the resulting clusters

In [103]:
import matplotlib.cm as cm
import matplotlib.colors as colors

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(northyork_merged['Latitude'], northyork_merged['Longitude'], northyork_merged['Neighbourhood'], northyork_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
    
map_clusters

### Examining Clusters

#### Cluster 1

In [104]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 0, northyork_merged.columns[[1] + list(range(5, northyork_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
0                                       Parkwoods                  Park   
1                                Victoria Village           Pizza Place   
2                 Lawrence Heights,Lawrence Manor        Clothing Store   
4                                       Glencairn                Arcade   
5                 Flemingdon Park,Don Mills South           Coffee Shop   
6                               Hillcrest Village           Golf Course   
7   Bathurst Manor,Downsview North,Wilson Heights           Coffee Shop   
8                      Fairview,Henry Farm,Oriole        Clothing Store   
9                  Northwood Park,York University           Coffee Shop   
13                                 Downsview West         Moving Target   
14               Downsview,North Park,Upwood Park                  Park   
17                              Downsview Central     Korean Restaurant   
18               Bedford Park,Lawrence Manor East  Fast Food Restaurant   
20                               Willowdale South      Ramen Restaurant   
21                            Downsview Northwest          Liquor Store   
22                                York Mills West                  Park   
23                                Willowdale West              Pharmacy   

         2nd Most Common Venue 3rd Most Common Venue  \
0                         Pool     Food & Drink Shop   
1                  Coffee Shop          Hockey Arena   
2       Furniture / Home Store         Women's Store   
4                       Bakery                   Pub   
5             Asian Restaurant                   Gym   
6                         Pool    Athletics & Sports   
7          Fried Chicken Joint         Shopping Mall   
8         Fast Food Restaurant           Coffee Shop   
9       Furniture / Home Store                   Bar   
13               Grocery Store                  Bank   
14  Construction & Landscaping                Bakery   
17                  Food Truck      Business Service   
18          Italian Restaurant           Coffee Shop   
20                 Coffee Shop        Sandwich Place   
21               Grocery Store  Gym / Fitness Center   
22           Electronics Store                  Bank   
23               Grocery Store           Pizza Place   

        4th Most Common Venue     5th Most Common Venue  \
0        Fast Food Restaurant            Discount Store   
1       Portuguese Restaurant             Women's Store   
2          Miscellaneous Shop       Arts & Crafts Store   
4         Japanese Restaurant             Women's Store   
5                  Beer Store                 Bike Shop   
6        Fast Food Restaurant  Mediterranean Restaurant   
7   Middle Eastern Restaurant        Frozen Yogurt Shop   
8                 Bus Station          Asian Restaurant   
9          Falafel Restaurant            Massage Studio   
13              Shopping Mall                   Dog Run   
14             Massage Studio                   Dog Run   
17             Baseball Field         Electronics Store   
18           Greek Restaurant         Indian Restaurant   
20           Sushi Restaurant                Restaurant   
21         Athletics & Sports            Discount Store   
22                Coffee Shop   Comfort Food Restaurant   
23                Coffee Shop                   Butcher   

         6th Most Common Venue       7th Most Common Venue  \
0               Clothing Store                 Coffee Shop   
1               Discount Store     Comfort Food Restaurant   
2                     Boutique                 Coffee Shop   
4            Electronics Store     Comfort Food Restaurant   
5               Clothing Store          Chinese Restaurant   
6                      Dog Run               Women's Store   
7                    Pet Store                    Pharmacy   
8                    Juice Bar                  Restaurant   
9                      Dog Run     Comfort Food Restaur

#### Cluster 2

In [105]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 1, northyork_merged.columns[[1] + list(range(5, northyork_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
11  CFB Toronto,Downsview East           Snack Place               Airport   
12     Silver Hills,York Mills                  Park             Cafeteria   

   3rd Most Common Venue 4th Most Common Venue    5th Most Common Venue  \
11                  Park     Electronics Store              Coffee Shop   
12               Dog Run           Coffee Shop  Comfort Food Restaurant   

         6th Most Common Venue       7th Most Common Venue  \
11     Comfort Food Restaurant  Construction & Landscaping   
12  Construction & Landscaping              Cosmetics Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
11        Cosmetics Shop         Deli / Bodega       Department Store  
12         Deli / Bodega      Department Store           Dessert Shop

#### Cluster 3

In [106]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 2, northyork_merged.columns[[1] + list(range(5, northyork_merged.shape[1]))]]

Neighbourhood       1st Most Common Venue 2nd Most Common Venue  \
19  Emery,Humberlea  Construction & Landscaping        Baseball Field   

   3rd Most Common Venue 4th Most Common Venue    5th Most Common Venue  \
19         Women's Store   Empanada Restaurant  Comfort Food Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19        Cosmetics Shop         Deli / Bodega      Department Store   

   9th Most Common Venue 10th Most Common Venue  
19          Dessert Shop     Dim Sum Restaurant

#### Cluster 4

In [107]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 3, northyork_merged.columns[[1] + list(range(5, northyork_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
15  Humber Summit           Pizza Place   Empanada Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
15               Dog Run        Clothing Store           Coffee Shop   

      6th Most Common Venue       7th Most Common Venue 8th Most Common Venue  \
15  Comfort Food Restaurant  Construction & Landscaping        Cosmetics Shop   

   9th Most Common Venue 10th Most Common Venue  
15         Deli / Bodega       Department Store

#### Cluster 5

In [108]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 5, northyork_merged.columns[[1] + list(range(5, northyork_merged.shape[1]))]]

Empty DataFrame
Columns: [Neighbourhood, 1st Most Common Venue, 2nd Most Common Venue, 3rd Most Common Venue, 4th Most Common Venue, 5th Most Common Venue, 6th Most Common Venue, 7th Most Common Venue, 8th Most Common Venue, 9th Most Common Venue, 10th Most Common Venue]
Index: []

#### No fifth cluster is present.